src/OpenTK.Mathematics/Geometry/Ray3.cs

```diff
@@ -0,0 +1,143 @@
+        public Vector3 Origin { get; set; }
+
+        private Vector3 _direction { get; set; }
+
+        /// <summary>
+        /// Direction of the ray.
+        /// </summary>
+        public Vector3 Direction {
+            get => _direction;
+            set => _direction = value.Normalized();
+        }
+
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) {
+            return obj is Ray3 ray && Equals(ray);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(Ray3 other) {
+            return Origin.Equals(other.Origin) &&
+                Direction.Equals(other.Direction);
+        }
+
+        /// <inheritdoc/>
+        public override int GetHashCode() {
+            return HashCode.Combine(Origin, Direction);
+        }
+
+        /// <inheritdoc/>
+        public override string ToString() {
+            return $"{Origin} -> {Direction}";
+        }
+
+        /// <param name="ray">The ray.</param>
+        /// <param name="distance">The distance to travel along the ray.</param>
+        /// <returns>The point <paramref name="distance"/> units along the ray. </returns>
+        [Pure]
+        public Vector3 GetPoint(float distance) {
+            return GetPoint(in this, distance);
+        }
+
+        /// <summary>
+        /// Get the closest point along the ray to a vector, with optional range clamping.
+        /// </summary>
+        /// <param name="ray">The ray</param>
+        /// <param name="vector">The vector</param>
+        /// <param name="nearClamp">Near clamp</param>
+        /// <param name="farClamp">Far clamp</param>
+        /// <returns>The closest point to the <paramref name="vector"/> along the <paramref name="ray"/>, within the clamped range.</returns>
+        [Pure]
+        public Vector3 GetClosestPoint(in Vector3 vector, float nearClamp = 0, float farClamp = float.PositiveInfinity) {
+            return GetClosestPoint(in this, in vector, nearClamp, farClamp);
+        }
+
+        /// <param name="ray">The ray.</param>
+        /// <param name="distance">The distance to travel along the ray.</param>
+        /// <returns>The point <paramref name="distance"/> units along the ray. </returns>
+        public static Vector3 GetPoint(in Ray3 ray, float distance) {
+            return ray.Origin + ray.Direction * distance;
+        }
+
+        /// <summary>
+        /// Get the closest point along the ray to a vector, with optional range clamping.
+        /// </summary>
+        /// <param name="ray">The ray</param>
+        /// <param name="vector">The vector</param>
+        /// <param name="nearClamp">Near clamp</param>
+        /// <param name="farClamp">Far clamp</param>
+        /// <returns>The closest point to the <paramref name="vector"/> along the <paramref name="ray"/>, within the clamped range.</returns>
+        public static Vector3 GetClosestPoint(in Ray3 ray, in Vector3 vector, float nearClamp = 0, float farClamp = float.PositiveInfinity) {
```

Comment:
I'm not a fan of the clamping tbh. Especially the `nearClamp` as we should in that case just add a `Line` struct as well as a `Ray` struct.
I think we should generally go for "rays have an origin and extend infinitely from that origin" as our base for thinking about rays (`t` values for intersection tests doesn't need to honor this as they often mean more than just some length from the origin).